src/github.com/matrix-org/dendrite/clientapi/readers/profile.go

```diff
@@ -175,13 +238,96 @@ func SetDisplayName(
 	}
 }
 
-func getLocalPart(userID string) string {
-	if !strings.HasPrefix(userID, "@") {
-		panic(fmt.Errorf("Invalid user ID"))
+func buildMembershipEvents(
+	memberships []authtypes.Membership, db *accounts.Database,
+	oldProfile *authtypes.Profile, changedKey string, newValue string,
+	userID string, cfg *config.Dendrite, queryAPI api.RoomserverQueryAPI,
+) ([]gomatrixserverlib.Event, error) {
+	ev := []gomatrixserverlib.Event{}
+
+	for _, membership := range memberships {
+		prevContent := events.MemberContent{
+			Membership:  "join",
+			DisplayName: oldProfile.DisplayName,
+			AvatarURL:   oldProfile.AvatarURL,
+		}
+
+		prev := prevMembership{
+			UserID:      userID,
+			PrevID:      membership.EventID,
+			PrevContent: prevContent,
+		}
+
+		builder := gomatrixserverlib.EventBuilder{
+			Sender:   userID,
+			RoomID:   membership.RoomID,
+			Type:     "m.room.member",
+			StateKey: &userID,
+		}
+
+		if err := builder.SetUnsigned(prev); err != nil {
+			return nil, err
+		}
```

Comment:
I don't think this is the right place to set the prevContent. Unfortunately the concept of a "prev_content" doesn't really make sense until the events have been linearised by the roomserver. So I think that in general this is something that has to be synthesised in the syncserver.